ArduCopter/mode_autorotate.cpp

```diff
@@ -0,0 +1,417 @@
+        }
+        //set position controller
+        pos_control->set_alt_target_from_climb_rate(_desired_v_z, G_Dt, false);
+
+        //update controllers
+        pos_control->update_z_controller();
+
+        //update bail out timer
+        bail_time_remain -= G_Dt;
+        
+        if (bail_time_remain <= 0.0f) {
+            //bail out timer complete.  Change flight mode.
+            set_mode(copter.prev_control_mode, MODE_REASON_AUTOROTATION_BAILOUT);
+        }
+
+        break;
+        }
+    }
+
+
+    switch (nav_pos_switch) {
+
+        case USER_CONTROL_STABILISED:
+        {
+            //Operator is in control of roll and yaw.  Controls act as if in stabilise flight mode.  Pitch is controlled 
+            //by speed-height controller.
+            float pilot_roll, pilot_pitch;
+            get_pilot_desired_lean_angles(pilot_roll, pilot_pitch, copter.aparm.angle_max, copter.aparm.angle_max);
+
+            //get pilot's desired yaw rate
+            float pilot_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+
+            //pitch target is calculated in autorotation phase switch above
+            attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(pilot_roll, _pitch_target, pilot_yaw_rate);
+            //attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(pilot_roll, pilot_pitch, pilot_yaw_rate);
+            break;
+        }
+
+        case STRAIGHT_AHEAD:
+        {
+            break;
+        }
+
+        case INTO_WIND:
+        {
+            break;
+        }
+
+        case NEAREST_RALLY:
+        {
+            break;
+        }
+    }
+
+    //Output warning messaged if rpm signal is bad
+    if (_flags.bad_rpm) {
+        warning_message(1);
+    }
+
+} //end function run()
+
+
+// Determine the glide path angle based on NED frame velocities.  This method determines the instaintanious glide angle,
+// not accounting for the wind.  This function is not yet used.  It will be adopted in the future for selection of most efficient
+// forward flight speed.
+float ModeAutorotate::get_ned_glide_angle(void)
+{
+    float sink_velocity = inertial_nav.get_velocity().z; //(cm/s)
+
+    //Convert sink velocity to m/s and convert sink velocity to be positive down
+    sink_velocity /= -100.0f; //(m/s)
+
+    //remove climb rates from sink velocity
+    if (sink_velocity < 0) {
+        sink_velocity = 0;
+    }
+
+    //Determine ground speed
+    Vector2f _groundspeed_vector = ahrs.groundspeed_vector();
+    float ground_speed_forward = _groundspeed_vector.x*ahrs.cos_yaw() + _groundspeed_vector.y*ahrs.sin_yaw(); //(m/s)
+
+    //catch divide by zero
+    if (ground_speed_forward < 0.01f) {
+        ground_speed_forward = 0.01f;
+    }
+
+    //Calculate glide slope
+    float windless_glide_slope = atanf(sink_velocity/ground_speed_forward);
+
+    return windless_glide_slope;
+}
+
+
+void ModeAutorotate::warning_message(uint8_t message_n)
```

Comment:
I guess it's to make things easier in the future but personally I'd remove this function and just add the send_texts in the single call to warning_message()

Reply:
I would like to leave this as is, if that is acceptable?  It does serve its purpose, albeit a simple one at the moment, and it will help to keep warning messages tidy and easy to find in my future work.  Multiple areas of the mode can call the same warning message so this is intended to avoid duplication.